libs/mpi/include/hpx/mpi/mpi_future.hpp

```diff
@@ -0,0 +1,186 @@
+    // An implementation of future_data for MPI
+    // -----------------------------------------------------------------
+    struct future_data : hpx::lcos::detail::future_data<void>
+    {
+        HPX_NON_COPYABLE(future_data);
+
+        using init_no_addref = hpx::lcos::detail::future_data<void>::init_no_addref;
+
+        // default empty constructor
+        future_data() = default;
+
+        // constructor that takes a request
+        future_data(init_no_addref no_addref, MPI_Request request)
+            : hpx::lcos::detail::future_data<void>(no_addref), request_(request)
+        {}
+
+        // constructor used for creation directly by invoke
+        future_data(init_no_addref no_addref)
+            : hpx::lcos::detail::future_data<void>(no_addref)
+        {}
+
+        // The native MPI request handle owned by this future data
+        MPI_Request request_;
+    };
+
+    // -----------------------------------------------------------------
+    // utility function to add a new request to the list to be tracked
+    void add_to_request_list(future_data_ptr data)
+    {
+        // lock the MPI work list before appending new request
+        std::unique_lock<std::mutex> lk(list_mtx_);
+        // debugging
+        // for debugging only
+        if (mpi_debug.is_enabled()) {
+            int rank = -1;
+            MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+            mpi_debug.debug(hpx::debug::str<>("add request")
+                , "R", rank
+                , "request", hpx::debug::hex<8>(data->request_)
+                , "list size", hpx::debug::dec<3>(active_requests_.size()+1));
+        }
+        // push onto list, this will make a copy and increment the ref count
+        active_requests_.push_back(data);
+    }
+
+    // -----------------------------------------------------------------
+    // return a future object from a user supplied MPI_Request
+    hpx::future<void> get_future(MPI_Request request)
+    {
+        // for debugging only
+        int rank = -1;
+        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+        if (request != MPI_REQUEST_NULL) {
+            future_data_ptr data =
+                new future_data(future_data::init_no_addref{}, request);
+
+            add_to_request_list(data);
+
+            using traits::future_access;
+            return future_access<hpx::future<void>>::create(std::move(data));
+        }
+        return hpx::make_ready_future<void>();
+    }
+
+    // -----------------------------------------------------------------
+    // return a future from an async call to MPI_Ixxx function
+    template <typename F, typename ...Ts>
+    hpx::future<void> async(F f, Ts &&...ts)
+    {
+        future_data_ptr data =
+            new future_data(future_data::init_no_addref{});
+
+        // invoke the call to MPI_Ixxx
+        f(std::forward<Ts>(ts)..., &data->request_);
+
+        // add the new shared state to the list for tracking
+        add_to_request_list(data);
```

Comment:
The PR can easily support `MPI_THREAD_SERIALIZED` by rewriting this to sth like:

```
{
        std::lock_guard<std::mutex> lk(list_mtx_);
        // invoke the call to MPI_Ixxx
        f(std::forward<Ts>(ts)..., &data->request_);

        // add the new shared state to the list for tracking
        add_to_request_list(data);
}
```

This ensures that non-blocking calls are serialized even if made from different threads.

Reply:
Should this lock be acquired unconditionally only to support `MPI_THREAD_SERIALIZED`?

Reply:
Also, I don't like the fact that we are using a `std::mutex` on a HPX thread. This could possibly stall the whole scheduler progress.

Reply:
We can perform a check at init time to see what level of thread support is available and then conditionally make use of a lock when invoking mpi functions.

Note that the polling function DOES happen on a std thread and not on an hpx thread and we use std::mutexes there (with a try_lock to avoid uneccessary waits). The sending happens on an HPX thread and we could use an hpx mutex - however, we wish to avoid high latencies on sends, so it might still be worth using a std::mutex. We don't want a thread suspension to kick in when we have an 'urgent' send, even an asynchronous one. I will look into options here. If the user has selected serialized threading in MPI, then they probably aren't too bothered by performance and we can use a lock...

Reply:
Fine with me. Note however that a `std::mutex` might incur much higher overheads than an hpx mutex.